heat/core/linalg/basics.py

```diff
@@ -1075,6 +1077,317 @@ def projection(a, b):
+        raise TypeError(f"out must be a ht.DNDarray or None not {type(out)}")
+
+    # ----------------------------------------------------------------------------
+    # ALGORITHM
+    # ----------------------------------------------------------------------------
+    # ---------------------------------------------
+    # CASE 2D input (ignore axis1, axis) => scalar
+    # ---------------------------------------------
+    if len(a.lshape) == 2:
+        # CASE 1.1: offset results into an empty array
+        if offset <= -a.gshape[0] or offset >= a.gshape[1]:
+            sum_along_diagonals_t = torch.tensor(
+                0, dtype=dtype.torch_type(), device=a.device.torch_device
+            )
+
+        # CASE 1.2: non-zero array, call torch.trace on concerned sub-DNDarray
+        else:
+            # determine the additional offset created by distribution of `a`
+            a_sub = a.copy()
+            if a.is_distributed():
+                offset_split, _, _ = a.comm.chunk(a.gshape, a.split)
+                if a.split == 0:
+                    offset += offset_split
+                # a.split == 1
+                else:
+                    offset -= offset_split
+
+            # Calculate resulting/concerned sub-array `a_sub`
+            if offset > 0:
+                offset = min(offset, a_sub.lshape[1])
+                a_sub = factories.array(
+                    a_sub.larray[:, offset:], device=a_sub.device, comm=a_sub.comm
+                )
+            elif offset < 0:
+                offset = min(-offset, a_sub.lshape[0])
+                a_sub = factories.array(
+                    a_sub.larray[offset:, :], device=a_sub.device, comm=a_sub.comm
+                )
+
+            # calculate trace /partial sum on that sub-array
+            if 0 not in a_sub.lshape:
+                sum_along_diagonals_t = torch.trace(a_sub.larray)
+
+                # make sure result is of correct dtype
+                sum_along_diagonals_t = sum_along_diagonals_t.type(a_sub.dtype.torch_type())
+
+            # empty array => result = 0
+            else:
+                sum_along_diagonals_t = torch.tensor(
+                    0, dtype=a_sub.dtype.torch_type(), device=a_sub.device.torch_device
+                )
+
+        if out is not None:
+            raise ValueError("`out` not applicable (/must be None) if result is a scalar/ input 2D")
+
+        # sum up all partial sums
+        if a.is_distributed():
+            a.comm.Allreduce(MPI.IN_PLACE, sum_along_diagonals_t, MPI.SUM)
+
+        # cast to DNDarray of correct dtype
+        sum_along_diagonals = factories.array(
+            sum_along_diagonals_t, split=a.split, dtype=dtype, device=a.device, comm=a.comm
+        )
+
+        # convert resulting 0-d DNDarray to scalar
+        return sum_along_diagonals.item()
+
+    # -------------------------------
+    # CASE > 2D => DNDArray
+    # -------------------------------
+    else:
```

Comment:
since the if block above ends with a return and the rest of the function is in an `else` block, this `else` can be removed and the code be one tab less